src/net/client/cache.rs

```diff
@@ -0,0 +1,1160 @@
+        }
+        target.push(rr).expect("push error");
+    }
+
+    let result = target.as_builder().clone();
+    Ok(
+        Message::<Bytes>::from_octets(result.finish().into_target().into())
+            .expect(
+                "Message should be able to parse output from MessageBuilder",
+            ),
+    )
+}
+
+/// Check if a type is a DNSSEC type that needs to be removed.
+fn is_dnssec(rtype: Rtype) -> bool {
+    rtype == Rtype::Rrsig || rtype == Rtype::Nsec || rtype == Rtype::Nsec3
+}
+
+/// This type represents that various subtypes of a NOERROR result.
+enum NoErrorType {
+    /// The result is an answer to the question.
+    Answer,
+
+    /// The name exists, but there is not data for the request class and tpye
+    /// combination.
+    NoData,
+
+    /// The upstream DNS server sent a delegation to another DNS zone.
+    Delegation,
+
+    /// None of the above. This is not a valid response.
+    NoErrorWeird,
+}
+
+/// Classify a responses with a NOERROR result.
+fn classify_no_error<Octs>(msg: &Message<Octs>) -> Result<NoErrorType, Error>
+where
+    Octs: Octets,
+{
+    // Check if we have something that resembles an answer.
+    let mut question_section = msg.question();
+    let question = question_section.next().expect("section expected")?;
+    let qtype = question.qtype();
+    let qclass = question.qclass();
+
+    // Note we only look qtype and qclass. The goal is not to perform
+    // a consistency check. Just whether there is supposed to be an
+    // answer or not.
+    let mut msg = msg.answer()?;
+    for rr in &mut msg {
+        let rr = rr?;
+        if rr.rtype() == qtype && rr.class() == qclass {
+            // We found an answer.
+            return Ok(NoErrorType::Answer);
+        }
+    }
+
+    // No answer. Check the authority section for SOA and NS records.
+    // If the SOA is present then the response is a NODATA response.
+    // If SOA records are absent but NS records are present then the
+    // response is a delegation.
+    let mut found_ns = false;
+    let mut msg = msg.next_section()?.expect("section should be present");
+    for rr in &mut msg {
+        let rr = rr?;
+        if rr.class() == qclass && rr.rtype() == Rtype::Soa {
+            return Ok(NoErrorType::NoData);
+        }
+        if rr.class() == qclass && rr.rtype() == Rtype::Ns {
+            found_ns = true;
+        }
+    }
+
+    if found_ns {
+        return Ok(NoErrorType::Delegation);
+    }
+
+    // Neither SOA nor NS were found. This is a broken response.
+    Ok(NoErrorType::NoErrorWeird)
+}
+
+/// Get the extended rcode of a message.
+fn get_opt_rcode<Octs: Octets>(msg: &Message<Octs>) -> OptRcode {
+    let opt = msg.opt();
+    match opt {
+        Some(opt) => opt.rcode(msg.header()),
+        None => {
+            // Convert Rcode to OptRcode, this should be part of
+            // OptRcode
+            OptRcode::from_int(msg.header().rcode().to_int() as u16)
+        }
+    }
```

Comment:
`Rcode` does support conversion to `OptRcode`.

You could alternately do one of the following:

```
    msg.opt().map_or_else(
        || msg.header().rcode().into(),
        |opt| opt.rcode(msg.header()),
    )
```

Or:

```
    msg.opt()
        .map(|opt| opt.rcode(msg.header()))
        .unwrap_or_else(|| msg.header().rcode().into())
```